library/src/main/scala-new-collections/scala/scalajs/js/ArrayOps.scala

```diff
@@ -8,97 +8,2002 @@
+     *
+     *  @param f
+     *    the function to apply to each element.
+     *  @tparam B
+     *    the element type of the returned array.
+     *  @return
+     *    a new aray resulting from applying the given function `f` to each
+     *    element of this array and collecting the results.
+     */
+    def map[B](f: A => B): js.Array[B] = {
+      val b = new js.Array[B]
+      val len = xs.length
+      var i = 0
+      while (i < len) {
+        val x = xs(i)
+        if (p(x))
+          b.push(f(x))
+        i = i + 1
+      }
+      b
+    }
+
+    /** Builds a new array by applying a function to all elements of this array
+     *  and using the elements of the resulting collections.
+     *
+     *  @param f
+     *    the function to apply to each element.
+     *  @tparam B
+     *    the element type of the returned array.
+     *  @return
+     *    a new array resulting from applying the given collection-valued
+     *    function `f` to each element of this array and concatenating the
+     *    results.
+     */
+    def flatMap[B](f: A => IterableOnce[B]): js.Array[B] = {
+      val b = new js.Array[B]
+      val len = xs.length
+      var i = 0
+      while (i < len) {
+        val x = xs(i)
+        if (p(x))
+          b.addAll(f(xs(i)))
+        i += 1
+      }
+      b
+    }
+
+    def flatMap[BS, B](f: A => BS)(
+        implicit asIterable: BS => scala.collection.Iterable[B]): js.Array[B] = {
+      flatMap[B](x => asIterable(f(x)))
+    }
+
+    /** Creates a new non-strict filter which combines this filter with the
+     *  given predicate.
+     */
+    def withFilter(q: A => Boolean): WithFilter[A] =
+      new WithFilter[A](a => p(a) && q(a), xs)
+  }
+
+  private class ArrayIterator[A](private[this] val xs: js.Array[A])
+      extends AbstractIterator[A] {
+
+    private[this] var pos = 0
+
+    def hasNext: Boolean = pos < xs.length
+
+    def next(): A = {
+      try {
+        val r = xs(pos)
+        pos += 1
+        r
+      } catch {
+        case _: ArrayIndexOutOfBoundsException =>
+          throw new NoSuchElementException
+      }
+    }
+  }
+
+  private class ReverseIterator[A](private[this] val xs: js.Array[A])
+      extends AbstractIterator[A] {
+
+    private[this] var pos = xs.length - 1
+
+    def hasNext: Boolean = pos >= 0
+
+    def next(): A = {
+      try {
+        val r = xs(pos)
+        pos -= 1
+        r
+      } catch {
+        case _: ArrayIndexOutOfBoundsException =>
+          throw new NoSuchElementException
+      }
+    }
+  }
+
+  private class GroupedIterator[A](xs: js.Array[A], groupSize: Int)
+      extends AbstractIterator[js.Array[A]] {
+
+    private[this] var pos = 0
+
+    def hasNext: Boolean = pos < xs.length
+
+    def next(): js.Array[A] = {
+      if (pos >= xs.length)
+        throw new NoSuchElementException
+      val r = new js.ArrayOps(xs).slice(pos, pos + groupSize)
+      pos += groupSize
+      r
+    }
```

Comment:
From my mails: @gzm0 says:

> Two things in general here:
>
> * I do not understand when something is marked inline and when it is not (e.g. iterator is not, but should?).
> * How do we ensure that new methods on collections also appear here?

Reply:
> I do not understand when something is marked inline and when it is not (e.g. iterator is not, but should?).

A method is `@inline` if at least one of the following applies:

* It was `@inline` in its implementation upstream
* It delegates to a JS operation
* It is of the form `{ someMethodCall(...); xs }`

Arguably, the last two bullets should be integrated as heuristics in the optimizer, but that's for another day.

> How do we ensure that new methods on collections also appear here?

We do as they do for `scala.collection.ArrayOps`: we monitor the `Seq`, `Buffer`, `Growable` and `Shrinkable` APIs, and we replicate the changes. It's sad, but the new collection design does not seem to allow anything else.

Reply:
Another thing in general: There are a lot of while loops in this file to replace `foreach` on `js.Array[_]`. From what I have understood, our optimizer should be able to inline `foreach` completely. Shouldn't we use the higher level construct then?